src/array-lib.jl

```diff
@@ -290,21 +290,45 @@ end
 @wrapped Base.map(f, x, y, z::AbstractArray, w...) = _map(f, x, y, z, w...)
 
 function _map(f, x, xs...)
+    return ArrayOp(
+        SymbolicUtils._promote_symtype(_map, (x,xs)),
+        (idx...,),
+        expr,
+        +,
+        Term{Any}(map, [f, x, xs...])
+    )
+end
+
+function SymbolicUtils._promote_symtype(::typeof(_map), args)
```

Comment:
I dont think we should be extending this function... It looks like a very internal helper function. Why can't we do this with just `promote_symtype`?

Reply:
That's what I attempted initially in 47801aae7f7f96c9a2d43bbf538b8099b38fc439 (btw sorry for the 3rd commit, did not know about `git rebase`)

I hope I remember correctly: The problem with `promote_symtype` for `_map` and `_mapreduce` is that `promote_symtype` is meant to have a signature `promote_symtype(f, arg_symtypes...)`. 
`_promote_symtype` receives the arguments instead of their types, i.e., `_promote_symtype(f, args...)`.
In case of `_map` or `_mapreduce` we would like to recursively call `promote_symtype` for the mapped function and maybe a binary operator.
With `promote_symtype` I tried to obtain function instances by accessing the corresponding field of the type: https://github.com/JuliaSymbolics/Symbolics.jl/blob/47801aae7f7f96c9a2d43bbf538b8099b38fc439/src/array-lib.jl#L314 Works only for singleton types, but that should still catch almost all use-cases. Feels awkward nonetheless.
With `_promote_symtype` and the arguments available, we can use exactly the same logic as if actually calling `_map` and `_mapreduce` to get consistent type predictions (another problem with the `promote_symtype` approach).